### ISSUE TITLE
Release

### ISSUE THREAD

**Issue by trthatcher** [trthatcher/MLKernels.jl#19]
@st-- 

I would _really_ like to finish up the non-derivative portion of the package ASAP (next 10 days or so) so that I may use it in my other projects and so that I can release a version of this package to http://pkg.julialang.org/. Originally, this would have happened last week, but the derivatives and ARD obviously added to that time. That's okay (:

I'm going through all the standard kernel definitions and correcting/refining all definitions to ensure they are clean and correct. I'm also going to revisit the standard kernel tests (broken on my last commit - don't worry, will fix up). The composite kernels will need to be sorted for the recursive definition.

Regarding the kernel derivatives, there's a couple options since they probably won't be ready. First, exclude them from the first release and just keep them in a development branch. Alternatively, they could be included in the package code, but not extracted/documented. You would need to explicitly import them. Lastly, they could be broken out into their own package with this one as a dependency (you would be the owner of the deriv package - but I could still co-maintain).

How would you like to approach it?

**Comment by st--** [comment]
Just fixed the kernel derivatives and tests. I changed the definition of the RationalQuadraticKernel so that beta only occurs once; you can still obtain the same kernels, you just have to adjust alpha differently - and now you can have parameter derivatives w.r.t. both, which is important for automatic parameter optimization.

Considering how closely linked a derivative is to the kernel, I think it'd be good to keep it in the same package: that makes it easier to check that, should you change the definition of a kernel, you also adjust the derivatives correspondingly. As far as I'm concerned the derivatives are pretty much ready - all the tests run through, so I'm pretty confident the implementation is correct. If there are no heavily-optimised versions of the kernelmatrix_dx/y/xy functions, so be it - but the generic fall-back functionality is there and working, so why not include it (export, documentation and all)?

Recursively composite kernels should, in principle, be working - I just haven't tested them thoroughly...

**Comment by st--** [comment]
For the release I think we should aim for 100% test coverage (not the coveralls value, because it claims some lines aren't tested even though they are, because they're inlined and I think only julia 0.4 can deal with that - also there are some lines which are marked tested but might not be, particularly the one-line tests for invalid parameter values, and any functions created through `@eval`!). Here is what's still missing in that regard:
- [ ] some type conversions, for the new ExponentialKernel and QuadraticKernel
- [ ] PeriodicKernel `description_string_long`
- [ ] ~~iscondposdef vs. isposdef (and if a kernel is pos.def., it's also cond.pos.def...)~~
- [ ] kernelparameters
- [ ] optimised kernelmatrix functions?
- [X] ~~ARD derivatives with respect to weights~~
- [ ] kernel matrix approximations

**Comment by trthatcher** [comment]
Luckily changing the definition of the RationalQuadratic kernel into that form worked out because ψ = log(1 + α‖x-y‖²) is negative definite and ϕ = exp(-βψ) is positive definite for negative definite ψ. I'm going to further generalize the kernel to (1 + α‖x-y‖²ᵞ)⁻ᵝ since the second derivative isn't so bad.

FYI, I'm defining simple versions of these kernels (eg. SquaredExponential) because any use of the `^` function calls an external library with some overhead (it seems like ‖x-y‖²ᵞ with γ = 1 takes ~50% longer than ‖x-y‖²), whereas `*` has very little effect on the run-time. I know it's extra work, but I feel it's worth it.

Agreed about 100% coverage. I'll address your points in order:
- [ ] I'll add these in as I complete the review of all the kernel definitions
- [ ] (see above)
- [ ] ~~`iscondposdef(kernel) = isposdef(kernel)` for standard kernels, I will define conditionally positive definite kernels explicitly.~~
- [ ] ~~What's missing for the `kernelparameters`?~~
- [ ] ~~I'd like to update the`kernelmatrix` functions myself for the recursive composite kernel definitions. I'm going to add a `!` version so that we can overwrite matrices.~~
- [ ] ~~Did you want to do the ARD derivatives?~~
- [ ] Nystrom approximation is finished for kernelmatrix(X,X), I need to look into kernelmatrix(X,Y)

**Comment by st--** [comment]
Sorry, my list was basically "this doesn't yet have a test for it" (and some of it, of course, also "wasn't implemented properly" - which the [lack of] test would point out)... so I think kernelparameters does all it needs to, just haven't written any tests yet.

**Comment by st--** [comment]
My current thoughts on open issues before releasing:
1. In 'master' you went back from 'ismercer' to 'isposdef_kernel', why? Does isposdef_kernel mean _strictly_ or _semi_ positive definite? If the former, it should just be Base.isposdef; if the latter, the method name should be more explicit - and I think 'ismercer' might be clearer there.
2. You removed the MercerSigmoidKernel from developments (but not master) - you wanted to throw it out because it's just a 'straightforward input transform' - I was thinking we could leave that as an explicit example/documentation, "this is how you can write a kernel which is just a transform"? And then people who want to use that can simply copy/include the example file...
3. I'll be happy to try and sort out the PeriodicKernel eventually, but I think we should move that out of master and leave it for a second release at some future date.
4. The nystrom() method needs to be fixed, thrown out, or replaced with the known working (but potentially slower?) version in test/test_approx.jl....

**Comment by trthatcher** [comment]
Comments on the above:
1. I have it as `ismercer` and `iscondposdef` in development. I was hoping for some kind of naming consistency, but that's not going to happen. However, I think it's important to include these tiers (mercer, conditionally positive definite and non-kernel)
2. Sure.
3. Sounds good - there's some considerations there. 
4. I imported a divide and conquer LAPACK function for symmetric Eigendecomposition which would be faster for inverting symmetric matrices than the SVD. I'm going to see if I can get it included in base Julia... I would say toss it for this release. There's a large run-time coefficient for the inversion so you need a small sample combined with a really large matrix to get any gain, anyway.

With respect to #33, my main concern was that the parameter derivatives have a bunch of nuances whereas the first and second partial derivatives of x and y are much more straightforward and predictable. I don't want to rush the parameter derivatives into the first release. Would you be open to approaching it like this:
- Include dx, dy and dxdy in the first release
- Create a default numerical version for `kappa_dz` and `kappa_d2z` for the two main kernel types. I think that if you define `kappa` - a single variable function - you should get the derivatives for free
- Return a "not implemented" sort of error by default - we can keep a chart of implemented routines for users to reference
- Defer `kernel_dp` to a later release to ensure a clean release (I'd like to be careful here with how the derivatives work and find a way to create default numerical derivatives because some derivatives are BAD)

Basically, I want to spend some extra time ensuring that the `kernel_dp` is clearly defined in terms of behavior and I don't want analytic derivatives to hold back the definition of new kernels when 90% of people will probably only need `kernelmatrix` (eg numerical fall-backs).

If this is an acceptable approach for you, #33 can be closed and we'll proceed as above.

**Comment by st--** [comment]
Sounds good to me!

**Comment by st--** [comment]
Merged into master and tidied up the code some more. I think it's ready, and we should release this as is, as 0.1. Then dx, dy, dxdy for 0.2 or something. But would be nice to get it into Pkg!

**Comment by trthatcher** [comment]
Awesome, thank you! That sounds great - I just want to take a quick look over the tests and then we should be good to go. dx, dy, dxdy should follow shortly after

**Comment by trthatcher** [comment]
I'll put the package in Pkg this evening.

**Comment by trthatcher** [comment]
@st-- We now have a package in Pkg (:

**Comment by Evizero** [comment]
Hi! This package is really cool! I'm very likely going to use it for [KSVM.jl](https://github.com/Evizero/KSVM.jl)
